test/test-tcp-write-in-a-row.c

```diff
@@ -0,0 +1,152 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static uv_tcp_t server;
+static uv_tcp_t client;
+static uv_tcp_t incoming;
+static int connect_cb_called;
+static int close_cb_called;
+static int connection_cb_called;
+static int write_cb_called;
+
+// 10 MB, which is large than the send buffer size and the recv buffer
+static char data[1024 * 1024 * 10];
+
+static void close_cb(uv_handle_t *handle)
+{
+  close_cb_called++;
+}
+
+static void write_cb(uv_write_t *w, int status)
+{
+  // the small write should finish immediately after the big write
+  ASSERT(uv_stream_get_write_queue_size((uv_stream_t *)&client) == 0);
+
+  write_cb_called++;
+
+  if (write_cb_called == 2) {
+    // we are done
+    uv_close((uv_handle_t *)&client, close_cb);
+    uv_close((uv_handle_t *)&incoming, close_cb);
+    uv_close((uv_handle_t *)&server, close_cb);
+  }
+
+  free(w);
+}
+
+static void connect_cb(uv_connect_t *_, int status)
+{
+  int r;
+  uv_buf_t buf;
+  uv_write_t *req;
+  size_t write_queue_size0, write_queue_size1;
+
+  ASSERT(status == 0);
+  connect_cb_called++;
+
+  req = (uv_write_t *)malloc(sizeof(uv_write_t));
+
+  // fire a big write
+  buf = uv_buf_init(data, sizeof(data));
+  r = uv_write(req, (uv_stream_t *)&client, &buf, 1, write_cb);
```

Comment:
Style: star leans left (here and elsewhere)
```suggestion
  r = uv_write(req, (uv_stream_t*)&client, &buf, 1, write_cb);
```